### PR TITLE
Replace synchronized by ReentrantLock

### PR DESCRIPTION
**Pull request by jbescos**
Synchronized blocks, will block platform threads using virtual threads with JDK 21+.

To avoid to block platform threads, we need to replace every synchronized by ReentrantLock.

This PR does not contain any other optimization.

Special attention reviewing:
`glassfish-hk2/hk2-runlevel/src/main/java/org/glassfish/hk2/runlevel/internal/CurrentTaskFuture.java`

I am not applying this change in tests and examples.

### CODE REVIEW

**Comment by jbescos** [comment]
It was not a trivial migration because HK2 uses `wait`, `notify` and `notifyAll`. Pay special attention reviewing that.

I already reviewed it, but it is good to put more eyes here.

Good news is that HK2 run a lot of concurrency tests that helped me with this change.

**Comment by dmatej** [comment]
I don't think the target branch should be a maintenance branch for 2.x, but rather the master for 3.1.0, which would be then merged into the 4.0.0-prepare. What do you think, @arjantijms .

**Comment by jbescos** [comment]
Hi @dmatej 

My PR comes here because Weblogic is having the version 2.6.0 (having javax) and it supports JDK 21 with virtual threads.

We can migrate these changes also to 3.x and higher.

**Comment by dmatej** [comment]
Ok, that makes sense to me.

**Comment by avpinchuk** [comment]
Hmm... Is calling `lock()` method inside try blocks a good practice?

**Comment by arjantijms** [comment]
> My PR comes here because Weblogic is having the version 2.6.0 (having javax) and it supports JDK 21 with virtual threads.

Ah, that explains it. Thanks!

(any plans for Weblogic to use the more recent 3.x version?)

**Comment by jbescos** [comment]
> > My PR comes here because Weblogic is having the version 2.6.0 (having javax) and it supports JDK 21 with virtual threads.
> 
> Ah, that explains it. Thanks!
> 
> (any plans for Weblogic to use the more recent 3.x version?)

I think I am not allowed to comment on this because Oracle didn't announce anything related to this. However, from my eyes, it makes sense to forward it to 3.x too :).


> Hmm... Is calling `lock()` method inside try blocks a good practice?

I found in the book `java concurrency in practice`, written by many authors of the java concurrency API, they do it in the way you say. However I cannot figure it out what could be the issue. Only if lock is null, it will rethrow again a null pointer exception in the finally block. Maybe this is enough reason to move it outside the `try`.

**Comment by avpinchuk** [comment]
> I found in the book `java concurrency in practice`, written by many authors of the java concurrency API, they do it in the way you say. However I cannot figure it out what could be the issue. Only if lock is null, it will rethrow again a null pointer exception in the finally block. Maybe this is enough reason to move it outside the `try`.

If `lock()` method throws an unchecked exception (albeit in very unlikely situations), then attemp to `unlock()` will cause `IllegalMonitorStateException` because the lock was not acquird.

**Comment by jbescos** [comment]
> > I found in the book `java concurrency in practice`, written by many authors of the java concurrency API, they do it in the way you say. However I cannot figure it out what could be the issue. Only if lock is null, it will rethrow again a null pointer exception in the finally block. Maybe this is enough reason to move it outside the `try`.
> 
> If `lock()` method throws an unchecked exception (albeit in very unlikely situations), then attemp to `unlock()` will cause `IllegalMonitorStateException` because the lock was not acquired.

Thanks, it makes sense to fix it. It is in place now.

**Comment by jbescos** [comment]
Luckily HK2 have tests for concurrency. I was running them in a loop because these errors are not deterministic with the next script:
```
#!/bin/bash
set -e

ROOT=~/workspace/glassfish-hk2
ITERATION=$((0))
cd $ROOT

while true
do
	ITERATION=$((ITERATION + 1))
        # Requires compilation every time tests are executed
	mvn clean install -DskipTests
	echo "*****************************************************************************************"
	echo "************************************** Iteration $ITERATION **************************************"
	echo "*****************************************************************************************"
	mvn test
```
And the next test failed after 25 tries:
```
[ERROR] Failures: 
[ERROR]   ListenerErrorTest.testHaltLevelRegressionOnErrorNoThreads:231
```
Then I did the same for the branch `2.x-maintenance`, and I also found the same error. So it looks the issue was still there.

**Comment by dmatej** [comment]
I have simplified the script to this:
```
count=2; while mvn clean test -pl :hk2-runlevel -Dtest=ListenerErrorTest && (( --count > 0 ));  do echo "SUCCEEDED!"; done;
```
In your case it failed on this line:
```
       Assert.assertTrue(levelFiveService.isPreDestroyCalled());
````
However until now I still wasn't able to reproduce the issue. I will let it to run thousand times ...

**Comment by dmatej** [comment]
Hmm, would it help if the field would be volatile? Also that might be a reason why it is not reproducible on some CPUs - if it is accessed by multiple threads.
https://github.com/eclipse-ee4j/glassfish-hk2/blob/ddea025a8dea71b59e2606bd8b60b6925e5c7a15/hk2-runlevel/src/test/java/org/glassfish/hk2/runlevel/tests/listener/LevelFiveService.java#L32

**Comment by avpinchuk** [comment]
> However until now I still wasn't able to reproduce the issue. I will let it to run thousand times ...

`ErrorTest.testErrorDown()` failed too at line 219 (Linux, core i3 10th gen)

**Comment by jbescos** [comment]
> Hmm, would it help if the field would be volatile? Also that might be a reason why it is not reproducible on some CPUs - if it is accessed by multiple threads.
> 
> https://github.com/eclipse-ee4j/glassfish-hk2/blob/ddea025a8dea71b59e2606bd8b60b6925e5c7a15/hk2-runlevel/src/test/java/org/glassfish/hk2/runlevel/tests/listener/LevelFiveService.java#L32

I have tried it, and still the same issue. I am going to check what is the test about.

**Comment by jbescos** [comment]
This is the stacktrace when `LevelFiveService#preDestroy` is invoked (in my branch [reentrantlock](https://github.com/jbescos/glassfish-hk2/tree/reentrantlock)). We have to check the classes involved in that. It seems when the issue happens, `LevelFiveService#preDestroy` is not invoked.

```
Daemon Thread [RunLevelControllerThread-1706009359520] (Suspended (breakpoint at line 39 in org.glassfish.hk2.runlevel.tests.listener.LevelFiveService))	
	org.glassfish.hk2.runlevel.tests.listener.LevelFiveService.preDestroy() line: 39	
	sun.reflect.NativeMethodAccessorImpl.invoke0(java.lang.reflect.Method, java.lang.Object, java.lang.Object[]) line: not available [native method]	
	sun.reflect.NativeMethodAccessorImpl.invoke(java.lang.Object, java.lang.Object[]) line: 62	
	sun.reflect.DelegatingMethodAccessorImpl.invoke(java.lang.Object, java.lang.Object[]) line: 43	
	java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object...) line: 498	
	org.glassfish.hk2.utilities.reflection.ReflectionHelper.invoke(java.lang.Object, java.lang.reflect.Method, java.lang.Object[], boolean) line: 1268	
	org.jvnet.hk2.internal.ClazzCreator<T>.preDestroyMe(T) line: 322	
	org.jvnet.hk2.internal.ClazzCreator<T>.dispose(T) line: 381	
	org.jvnet.hk2.internal.AutoActiveDescriptor<T>.dispose(T) line: 156	
	org.jvnet.hk2.internal.SystemDescriptor<T>.dispose(T) line: 525	
	org.glassfish.hk2.runlevel.internal.AsyncRunLevelContext.destroyOne(org.glassfish.hk2.api.ActiveDescriptor<?>) line: 422	
	org.glassfish.hk2.runlevel.RunLevelContext.destroyOne(org.glassfish.hk2.api.ActiveDescriptor<?>) line: 81	
	org.jvnet.hk2.internal.ServiceHandleImpl<T>.close() line: 176	
	org.jvnet.hk2.internal.ServiceHandleImpl<T>(org.glassfish.hk2.api.ServiceHandle<T>).destroy() line: 67	
	org.glassfish.hk2.runlevel.internal.CurrentTaskFuture$DownQueueRunner.run() line: 1428	
	java.util.concurrent.ThreadPoolExecutor.runWorker(java.util.concurrent.ThreadPoolExecutor$Worker) line: 1149	
	java.util.concurrent.ThreadPoolExecutor$Worker.run() line: 624	
	org.glassfish.hk2.runlevel.internal.AsyncRunLevelContext$RunLevelControllerThread(java.lang.Thread).run() line: 748	
```

**Comment by jbescos** [comment]
I am making a new commit with the fix of ListenerErrorTest. I am not able to get any error now.

`12th Gen Intel(R) Core(TM) i7-1270P`

**Comment by arjantijms** [comment]
Running it 2000 times in parallel (from two checkouts), 4.0.0-prepare branch, JDK 21, macOS, M1 CPU

In 1 shell it failed after a while with:

```
[INFO] Running org.glassfish.hk2.runlevel.tests.listener.ListenerErrorTest
[ERROR] Tests run: 11, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.128 s <<< FAILURE! -- in org.glassfish.hk2.runlevel.tests.listener.ListenerErrorTest
[ERROR] org.glassfish.hk2.runlevel.tests.listener.ListenerErrorTest.testHaltLevelRegressionOnErrorNoThreads -- Time elapsed: 0.005 s <<< FAILURE!
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:87)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.junit.Assert.assertTrue(Assert.java:53)
	at org.glassfish.hk2.runlevel.tests.listener.ListenerErrorTest.testHaltLevelRegressionOnErrorNoThreads(ListenerErrorTest.java:231)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:316)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:240)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:214)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:155)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ListenerErrorTest.testHaltLevelRegressionOnErrorNoThreads:231
[INFO] 
[ERROR] Tests run: 11, Failures: 1, Errors: 0, Skipped: 0
```

The other shell is still going atm without any error.

**Comment by jbescos** [comment]
> Running it 2000 times in parallel (from two checkouts), 4.0.0-prepare branch, JDK 21, macOS, M1 CPU
> 
> ```
> java -version
> openjdk version "21.0.1" 2023-10-17 LTS
> OpenJDK Runtime Environment Temurin-21.0.1+12 (build 21.0.1+12-LTS)
> OpenJDK 64-Bit Server VM Temurin-21.0.1+12 (build 21.0.1+12-LTS, mixed mode)
> ```
> 
> In 1 shell it failed after a while with:
> 
> ```
> [INFO] Running org.glassfish.hk2.runlevel.tests.listener.ListenerErrorTest
> [ERROR] Tests run: 11, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.128 s <<< FAILURE! -- in org.glassfish.hk2.runlevel.tests.listener.ListenerErrorTest
> [ERROR] org.glassfish.hk2.runlevel.tests.listener.ListenerErrorTest.testHaltLevelRegressionOnErrorNoThreads -- Time elapsed: 0.005 s <<< FAILURE!
> java.lang.AssertionError
> 	at org.junit.Assert.fail(Assert.java:87)
> 	at org.junit.Assert.assertTrue(Assert.java:42)
> 	at org.junit.Assert.assertTrue(Assert.java:53)
> 	at org.glassfish.hk2.runlevel.tests.listener.ListenerErrorTest.testHaltLevelRegressionOnErrorNoThreads(ListenerErrorTest.java:231)
> 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
> 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
> 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
> 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
> 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
> 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
> 	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
> 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
> 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
> 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
> 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
> 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
> 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
> 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
> 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
> 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
> 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:316)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:240)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:214)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:155)
> 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
> 	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
> 	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
> 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
> 
> [INFO] 
> [INFO] Results:
> [INFO] 
> [ERROR] Failures: 
> [ERROR]   ListenerErrorTest.testHaltLevelRegressionOnErrorNoThreads:231
> [INFO] 
> [ERROR] Tests run: 11, Failures: 1, Errors: 0, Skipped: 0
> ```
> 
> The other shell is still going atm without any error. Update: it took hundreds of runs, but I did get the failure again (twice even).

Did you try it with my latest commit?

**Comment by dmatej** [comment]
Hurray, I reproduced the issue on eclipse/2.x-maintenance branch after 388 iterations too! :)
I will try your branch now.

**Comment by avpinchuk** [comment]
So, the problem not in tests but in HK2 itself.

 I'll do extensive testing this weekend and then prepare the patch.

**Comment by jbescos** [comment]
> So, the problem not in tests but in HK2 itself.
> 
> I'll do extensive testing this weekend and then prepare the patch.

In my branch it is passing. The issue is in the test, that it is checking one boolean before is set. Here is the fix of my commit: [commit](https://github.com/eclipse-ee4j/glassfish-hk2/pull/963/commits/07a0607ad11e4a09ab946e0973ad0ac26901c207)

**Comment by avpinchuk** [comment]
@jbescos, [here](https://github.com/eclipse-ee4j/glassfish-hk2/pull/970/commits/90eabf63fa6ed8ad25d27d647b9814f512d423e6) fix for `master` branch.

**Comment by jbescos** [comment]
> @jbescos, [here](https://github.com/eclipse-ee4j/glassfish-hk2/pull/970/commits/90eabf63fa6ed8ad25d27d647b9814f512d423e6) fix for `master` branch.

Thank you, I have added it in my latest commit.

**Review by dmatej — APPROVED** [review]

File: hk2-runlevel/src/test/java/org/glassfish/hk2/runlevel/tests/listener/LevelFiveService.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2013, 2018 Oracle and/or its affiliates. All rights reserved.
```
Comment (dmatej) [inline hk2-runlevel/src/test/java/org/glassfish/hk2/runlevel/tests/listener/LevelFiveService.java:2]: ... and in the LevelFiceService too

File: hk2-runlevel/src/test/java/org/glassfish/hk2/runlevel/tests/listener/ListenerErrorTest.java [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
```
[the remark this answers is not in the document]
Comment (jbescos) [reply hk2-runlevel/src/test/java/org/glassfish/hk2/runlevel/tests/listener/ListenerErrorTest.java]: It is reverted now

**Review by dmatej — APPROVED** [review]

**Review by senivam — APPROVED** [review]

**Pull request merged**